hw/ip/otbn/util/check_loop.py

```diff
@@ -0,0 +1,250 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import sys
+from typing import Dict, List, Tuple
+
+from shared.check import CheckResult
+from shared.decode import OTBNProgram, decode_elf
+from shared.insn_yaml import Insn
+
+
+class CodeSection:
+    '''A continuous part of a program's code (represented as a PC range).
+
+    The code section is considered to include both the start and end PC.
+    '''
+    def __init__(self, start: int, end: int):
+        self.start = start
+        self.end = end
+
+    def __contains__(self, pc):
+        return self.start <= pc and pc <= self.end
+
+    def __repr__(self):
+        return '0x{:x}-0x{:x}'.format(self.start, self.end)
+
+
+def _get_pcs_for_mnemonic(program: OTBNProgram, mnem: str) -> List[int]:
+    '''Gets all PCs in the program at which the given instruction is present.'''
+    imem_start = program.get_pc_at_symbol('_imem_start')
+    imem_end = program.get_pc_at_symbol('_imem_end')
+    out = []
+    for pc in range(imem_start, imem_end, 4):
+        insn = program.get_insn(pc)
+        if insn.mnemonic == mnem:
+            out.append(pc)
+    return out
+
+
+def _get_branches(program: OTBNProgram) -> List[int]:
+    '''Gets the PCs of all branch instructions (BEQ and BNE) in the program.'''
+    return _get_pcs_for_mnemonic(program, 'bne') + _get_pcs_for_mnemonic(
+        program, 'beq')
+
+
+def _get_loop_starts(program: OTBNProgram) -> List[int]:
+    '''Gets the start PCs of all loops (LOOP and LOOPI) in the program.'''
+    return _get_pcs_for_mnemonic(program, 'loop') + _get_pcs_for_mnemonic(
+        program, 'loopi')
+
+
+def _get_loops(program: OTBNProgram) -> List[CodeSection]:
+    '''Gets the PC ranges of all loops (LOOP and LOOPI) in the program.'''
+    loop_starts = _get_loop_starts(program)
+    loops = []
+    for pc in loop_starts:
+        insn = program.get_insn(pc)
+        operands = program.get_operands(pc)
+        end_pc = pc + operands['bodysize'] * 4
+        loops.append(CodeSection(pc + 4, end_pc))
+    return loops
+
+
+def _check_loop_iterations(program: OTBNProgram,
+                           loops: List[CodeSection]) -> CheckResult:
+    '''Checks number of iterations for loopi.
+
+    If the number of iterations is 0, this check fails; `loopi` requires at least
+    one iteration and will raise a LOOP error otherwise. The `loop` instruction
```

Comment:
I'm surprised this even assembles.

Reply:
This is representable in the ISA to match the behaviour of LOOP. You want LOOP to take the count directly (because otherwise people will be messing around subtracting one all over the place) but then it's nice to be able to grab bits directly from the encoding for LOOPI and shove them into the same logic as used for the register contents for LOOP.

We *could* make it an error in the assembler, as you say. But we'd need a "no, I really mean it" mode for DV (where we want to generate all sorts of naughty things) and we'd probably think of all sorts of other things that we could statically check... which is what Jade is building!

Reply:
My general idea here (or, certainly, my aspiration) is to keep the assembler/linker very small, since they're trusted to generate the binaries and we really don't want problems there, and then have a suite of untrusted post-checks on the generated binaries that (possibly over-conservatively) try to catch issues.